File: practice/data_structures/linked_lists/insert_node_at_a_specific_position_in_linked_list/insert_node_at_a_specific_position_in_linked_list.cpp
/*
  Insert Node at a given position in a linked list 
  head can be NULL 
  First element in the linked list is at position 0
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/

Node * InsertNth(Node * i_head, int i_data, int i_position)
{
    Node * node = new Node;
    node->data = i_data;
    
    if (i_position == 0)
    {
        node->next = i_head;
        return node;
    }
    
    Node * head = i_head;
    while (i_position > 1)
    {
        head = head->next;
        i_position--;
    }
    
    node->next = head->next;
    head->next = node;
    
    return i_head;
}

File: practice/data_structures/linked_lists/insert_node_at_a_specific_position_in_linked_list/insert_node_at_a_specific_position_in_linked_list.py
"""
 Insert Node at the end of a linked list
 head input could be None as well for empty list
 Node is defined as
 
 class Node(object):
 
   def __init__(self, data=None, next_node=None):
       self.data = data
       self.next = next_node

 return back the head of the linked list in the below method. 
"""

def InsertNth(i_head, i_data, i_position):
    if i_position == 0:
        return Node(i_data, i_head)
    cp = i_head
    while i_position > 1:
        cp = cp.next
        i_position -= 1
    n = Node(i_data, cp.next)
    cp.next = n
    return i_head

